File: frontend/src/lib/payable.ts
// 应付款前端类型 + API。对齐后端 schemas/payable.py。
// 🔴 整域红线:仅持 payable:read 者可见;后端整端点/整块门控。前端不自行判权限藏真值,
// 只在响应含 payable 键 / /payables 返回 200 时渲染。
import { api } from "./api";
import type { Page } from "./catalog";
import { formatMoney } from "./format";
import { qs } from "./qs";

/** 派生状态:未付 / 部分付 / 已付清(后端由 amount_* 单一口径派生)。 */
export type PayableStatus = "UNPAID" | "PARTIALLY_PAID" | "PAID";

/** 入库详情内嵌 payable 块(有活动应付且持 payable:read 时下发)。 */
export interface PayableOut {
  id: number;
  inbound_order_id: number;
  purchase_order_id: number;
  supplier_id: number;
  currency: string;
  amount_original: number | string;
  amount_credited: number | string;
  amount_allocated: number | string;
  amount_outstanding: number | string;
  status: PayableStatus;
  due_at: string | null;
  created_at: string;
}

export interface PayableListItem {
  id: number;
  inbound_order_id: number;
  inbound_order_no: string;
  purchase_order_id: number;
  purchase_order_no: string;
  supplier_id: number;
  supplier_display: string;
  currency: string;
  amount_original: number | string;
  amount_credited: number | string;
  amount_allocated: number | string;
  amount_outstanding: number | string;
  status: PayableStatus;
  due_at: string | null;
  created_at: string;
  /** 该供应商名下有未分配付款(可一键核销此账)。 */
  counterparty_has_unallocated: boolean;
}

/** 应付详情内活动核销记录行:哪笔付款、冲了多少、何时。 */
export interface PayableAllocationRow {
  id: number;
  payment_id: number;
  payment_no: string;
  amount: number;
  alloc_type: "AUTO" | "MANUAL";
  created_at: string;
}

/** 应付详情(账头 + 活动核销记录),GET /payables/{id}。🔴红线。 */
export interface PayableDetail {
  id: number;
  inbound_order_id: number;
  inbound_order_no: string;
  purchase_order_id: number;
  supplier_id: number;
  supplier_display: string | null;
  currency: string;
  amount_original: number;
  amount_credited: number;
  amount_allocated: number;
  amount_outstanding: number;
  status: PayableStatus;
  due_at: string | null;
  created_at: string;
  allocations: PayableAllocationRow[];
}

/** 派生状态徽标映射。 */
export const PAYABLE_STATUS_META: Record<PayableStatus, { label: string; color: string }> = {
  UNPAID: { label: "未付", color: "warning" },
  PARTIALLY_PAID: { label: "部分付", color: "processing" },
  PAID: { label: "已付清", color: "success" },
};

/** 金额格式化(应付域内均为真值,不涉红线脱敏)。 */
export function formatAmount(v: number | string): string {
  return formatMoney(v);
}

export interface PayableListFilters {
  supplier_id?: number;
  currency?: string;
  /** 派生状态筛选(服务端谓词镜像 derive_payable_status)。 */
  status?: PayableStatus;
  /** 入库单号 / 采购单号 / 供应商名 模糊搜索。 */
  q?: string;
  page?: number;
  size?: number;
}

export const payableApi = {
  list: (p: PayableListFilters) =>
    api.get<Page<PayableListItem>>(`/api/v1/payables${qs(p as Record<string, unknown>)}`),
  get: (id: number) => api.get<PayableDetail>(`/api/v1/payables/${id}`),
};

File: backend/app/schemas/payable.py
"""应付款 schemas。

🔴 整表红线域:仅在持 payable:read 的端点/块出现,不做字段级脱敏(整端点/整块即红线门)。
status(未付/部分付/已付清)由 amount_* 派生输出(derive_payable_status,单一口径)。
"""
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel

from app.db.models.payable import derive_payable_status


class PayableOut(BaseModel):
    """入库详情 payable 块(有活动应付且持 payable:read 时下发)。"""
    id: int
    inbound_order_id: int
    purchase_order_id: int
    supplier_id: int
    currency: str
    amount_original: float
    amount_allocated: float
    balance: float
    status: str
    due_at: date | None
    created_at: datetime

    @classmethod
    def build(cls, p) -> dict:
        return {
            "id": p.id, "inbound_order_id": p.inbound_order_id,
            "purchase_order_id": p.purchase_order_id, "supplier_id": p.supplier_id,
            "currency": p.currency, "amount_original": float(p.amount_original),
            "amount_allocated": float(p.amount_allocated), "balance": float(p.balance),
            "status": derive_payable_status(p.amount_original, p.amount_allocated),
            "due_at": p.due_at, "created_at": p.created_at,
        }


class PayableListItem(BaseModel):
    id: int
    inbound_order_id: int
    inbound_order_no: str
    purchase_order_id: int
    purchase_order_no: str
    supplier_id: int
    supplier_display: str
    currency: str
    amount_original: float
    amount_allocated: float
    balance: float
    status: str
    due_at: date | None
    created_at: datetime
    # D10:该供应商是否有未分配付款余额(预付),供列表提示「一键核销」入口(纯提示,非自动)。
    counterparty_has_unallocated: bool = False

    @classmethod
    def build(cls, item: dict) -> dict:
        return {
            **{k: item[k] for k in (
                "id", "inbound_order_id", "inbound_order_no", "purchase_order_id",
                "purchase_order_no", "supplier_id", "supplier_display", "currency", "due_at",
                "created_at")},
            "amount_original": float(item["amount_original"]),
            "amount_allocated": float(item["amount_allocated"]),
            "balance": float(item["balance"]),
            "status": derive_payable_status(item["amount_original"], item["amount_allocated"]),
            "counterparty_has_unallocated": item.get("counterparty_has_unallocated", False),
        }
